src/rust/engine/src/nodes.rs

```diff
@@ -644,6 +659,84 @@ impl From<Snapshot> for NodeKey {
   }
 }
 
+// TODO: Cache these downloads somewhere on-disk (like in ~/.cache/pants), and read them in some trusted way
+// TODO: Retry failures
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct DownloadedFile(Key);
+
+impl WrappedNode for DownloadedFile {
+  type Item = Arc<fs::Snapshot>;
+
+  fn run(self, context: Context) -> NodeFuture<Arc<fs::Snapshot>> {
+    let url_to_fetch = externs::project_str(&externs::val_for(&self.0), "url");
+
+    // Parse URL, get file
+    let url = try_future!(
+      Url::parse(&url_to_fetch).map_err(|err| throw(&format!("Error parsing URL: {}", err)))
+    );
+    let file_name = url
+      .path_segments()
+      .expect("TODO")
```

Comment:
Should fix before landing. Multiple `expects` like this seem like a good situation for combinators like `map`/`and_then`:
```
let file_name = if let Some(..) = url.path_segments().and_then(|ps| ps.last()) {
  ..
} else {
  // return a Future::Err
};
```